### PR TITLE
eye operator, for default storage type

### PR DESCRIPTION
**Pull request by ZiyueHuang**
## Description ##
eye operator, for default storage type

Requested in https://discuss.mxnet.io/t/is-there-an-eye-function-in-the-ndarray-api/526/5

cc @eric-haibin-lin 

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] eye, unittest

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: tests/python/unittest/test_ndarray.py [hunk 1]
```diff
@@ -736,6 +736,14 @@ def test_output():
     assert_almost_equal(out.asnumpy(), ones.asnumpy() * 2)
     arange_out = mx.nd.arange(0, 20, dtype='int64')
     assert_almost_equal(arange_out.asnumpy(), np.arange(0, 20))
+    N_array = np.random.randint(1, high=3, size=3)
+    M_array = np.random.randint(1, high=3, size=3)
+    k_array = np.random.randint(-5, high=5, size=3)
```
Comment (piiswrong) [inline tests/python/unittest/test_ndarray.py]: Could you run some more trials?
At least locally & with larger ranges. Just to verify if there are edge cases

Comment (piiswrong) [reply tests/python/unittest/test_ndarray.py]: Also I think there is the case where M is 0

File: src/operator/tensor/init_op.h [hunk 2]
```diff
@@ -63,6 +63,86 @@ struct InitOpParam : public dmlc::Parameter<InitOpParam> {
...
+  SHAPE_ASSIGN_CHECK(*out_attrs, 0, mshadow::Shape2(param.N, param.M > 0 ? param.M : param.N));
+  return true;
+}
+
+template<int req>
+struct eye_dns_fill {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, DType* out_data, const nnvm::dim_t num_cols,
+                                  const nnvm::dim_t k) {
+    if ((i % num_cols) == ((i / num_cols) + k)) {
```
Comment (cjolivier01) [inline src/operator/tensor/init_op.h]: Looks like two divides per value-fill. Would it be faster ( at least for CPU), to fill with 0's and then "walk" across an offset (using only add) to fill in the one's?

**Comment by piiswrong** [comment]
@ZiyueHuang ping

**Comment by ZiyueHuang** [comment]
Thanks for your comments. Sorry for the late response.

@piiswrong I increased the size and the number of trials and added M=0 case in unittest. I also run 1000 times locally(ubuntu).

@cjolivier01 I changed to two kernels. About 10 times faster on cpu.

File: src/operator/tensor/init_op.h [hunk 3]
```diff
@@ -428,6 +405,39 @@ void FillComputeZerosEx(const nnvm::NodeAttrs& attrs,
...
+  const nnvm::dim_t nnz = param.k > 0 ? std::min(cnnz, param.N) :
+                                        std::min(rnnz, num_cols);
+  using namespace mxnet_op;
+  MSHADOW_TYPE_SWITCH(out_data.type_flag_, DType, {
+    MXNET_ASSIGN_REQ_SWITCH(req[0], req_type, {
+      Fill(s, out_data, req[0], static_cast<DType>(0));
+      if (nnz > 0) {
+        Kernel<eye_dns_fill<req_type>, xpu>::Launch(
+          s, nnz, out_data.dptr<DType>(),
+          std::max((nnvm::dim_t)0, param.k), param.k, num_cols);
```
Comment (cjolivier01) [inline src/operator/tensor/init_op.h]: nit: 
```cpp
static_cast<nnvm::dim_t>(0)
```
or
```cpp
0LL
```
although the latter sort of ties dim_t to a specific type.

**Pull request merged**